pkg/dev/dev.go

```diff
@@ -53,8 +59,49 @@ func (o *DevClient) Start(devfileObj parser.DevfileObj, platformContext kubernet
 	if err != nil {
 		return err
 	}
-	klog.V(4).Infoln("Successfully created inner-loop resourcs")
-	fmt.Fprintf(out, "\nYour application is running on cluster.\n ")
+	klog.V(4).Infoln("Successfully created inner-loop resources")
+
+	return nil
+}
+
+func (o *DevClient) Cleanup() error {
+	var err error
+	return err
+}
+
+func (o *DevClient) SetupPortForwarding(pod *corev1.Pod, portPairs []string, errOut io.Writer) error {
+	transport, upgrader, err := spdy.RoundTripperFor(o.kubernetesClient.GetClientConfig())
+	if err != nil {
+		return err
+	}
+
+	req := o.kubernetesClient.GeneratePortForwardReq(pod.Name)
+
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, "POST", req.URL())
+	stopChan := make(chan struct{}, 1)
+	// passing nil for readyChan because it's eventually being closed if it's not nil
+	// passing nil for out because we only care for error, not for output messages; we want to print our own messages
+	fw, err := portforward.NewOnAddresses(dialer, []string{"localhost"}, portPairs, stopChan, nil, nil, errOut)
+	if err != nil {
+		return err
+	}
+
+	// start port-forwarding
+	err = fw.ForwardPorts()
+	if err != nil {
+		fmt.Fprint(errOut, fmt.Errorf("error setting up port forwarding: %v", err).Error())
```

Comment:
Why not returning an error here with the message?
If you want a different behaviour in the caller depending on the type of error, you could return some specific error type and test for it

Reply:
Returning the error makes sense. This part hasn't changed since the first iterations I did before organizing the code.